### PR TITLE
Improve certificate verification

### PR DESCRIPTION
**Pull request by arturkow2000**

### CODE REVIEW

File: src/certmgr/verify.rs [hunk 1]
```diff
@@ -128,13 +141,72 @@ impl CertMgr {
...
+        }
+
+        // Subject Alternative Name extension must be present if subject is
+        // empty.
+        // TODO: DN parsing must be updated, currently unknown OIDs in DN are
+        // discarded, so we can get empty DN where it actually isn't.
+        //
+        // Currently we are not using it alternative name so we just ignore it.
+        //
+        // Subject alternative name contains various intereseting
```
Comment (krystian-hebel) [inline src/certmgr/verify.rs]:
```suggestion
        // Subject alternative name contains various interesting
```

File: src/certmgr/verify.rs [hunk 2]
```diff
@@ -128,13 +141,72 @@ impl CertMgr {
...
+            error!("Expected X509v3 certificate but got v{}", cert.version());
+            return false;
+        }
+
+        // Subject Alternative Name extension must be present if subject is
+        // empty.
+        // TODO: DN parsing must be updated, currently unknown OIDs in DN are
+        // discarded, so we can get empty DN where it actually isn't.
+        //
+        // Currently we are not using it alternative name so we just ignore it.
```
Comment (krystian-hebel) [inline src/certmgr/verify.rs]:
```suggestion
        // Currently we are not using alternative name so we just ignore it.
```

File: src/certmgr/verify.rs [hunk 3]
```diff
@@ -128,13 +141,72 @@ impl CertMgr {
...
+        //
+        // Subject alternative name contains various intereseting
+        // information like TPM manufacturer, TPM model, TPM version and other
+        // TODO: do we need anything from there?
+
+        // EK certficate must have Basic Constraints with CA set to FALSE
+
+        if Self::ca_constaint_check(cert).0 {
```
Comment (krystian-hebel) [inline src/certmgr/verify.rs]:
```suggestion
        if Self::ca_constraint_check(cert).0 {
```

File: src/certmgr/verify.rs [hunk 4]
```diff
@@ -148,7 +220,7 @@ impl CertMgr {
 
     /// Checks whether certificate can be used for signing other certificates
     /// (using X.509v3 Constraints extension).
-    fn ca_constaint_check(cert: &X509Certificate) -> bool {
+    fn ca_constaint_check(cert: &X509Certificate) -> (bool, Option<u8>) {
```
Comment (krystian-hebel) [inline src/certmgr/verify.rs]:
```suggestion
    fn ca_constraint_check(cert: &X509Certificate) -> (bool, Option<u8>) {
```

File: src/certmgr/verify.rs [hunk 5]
```diff
@@ -197,21 +266,29 @@ impl CertMgr {
...
+        depth: usize,
     ) -> Result<bool>
     where
         T: trussed::client::Sha256,
     {
         // TODO: verify whether child issuer matches with parent subject.
 
         let parent_key = if let Some(parent) = parent {
-            if !Self::ca_constaint_check(parent) {
+            let (ca, path_len_constraint) = Self::ca_constaint_check(parent);
```
Comment (krystian-hebel) [inline src/certmgr/verify.rs]:
```suggestion
            let (ca, path_len_constraint) = Self::ca_constraint_check(parent);
```

File: src/certmgr/verify.rs [hunk 6]
```diff
@@ -197,21 +266,29 @@ impl CertMgr {
...
+            if let Some(path_len_constraint) = path_len_constraint.map(usize::from) {
+                if depth > path_len_constraint {
+                    return Err(Error::ExceededPathLenConstraint);
+                }
+            }
+
             parent.key()?
         } else {
-            if !Self::ca_constaint_check(child) {
+            if !Self::ca_constaint_check(child).0 {
```
Comment (krystian-hebel) [inline src/certmgr/verify.rs]:
```suggestion
            if !Self::ca_constraint_check(child).0 {
```

**Review by krystian-hebel — COMMENTED** [review]
Just few typos

File: src/certmgr/verify.rs (hunk 1 above)
Comment (arturkow2000) [reply src/certmgr/verify.rs]: Fixed

File: src/certmgr/verify.rs (hunk 2 above)
Comment (arturkow2000) [reply src/certmgr/verify.rs]: Fixed

**Review by krystian-hebel — APPROVED** [review]

**Pull request merged**